src/libraries/Microsoft.PowerFx.Core/Types/DType.cs

```diff
@@ -1564,6 +1564,75 @@ public DType DropMulti(ref bool fError, DPath path, params DName[] rgname)
             return SetType(ref fError, path, new DType(typeOuter.Kind, tree, AssociatedDataSources, DisplayNameProvider));
         }
 
+        public bool ContainsKindExpanded(DPath path, DKind kind)
+        {
+            AssertValid();
+            Contracts.Assert(kind >= DKind._Min && kind < DKind._Lim);
+
+            var fullType = this;
+            if (IsLazyType)
+            {
+                fullType = LazyTypeProvider.GetExpandedType(IsTable);
+            }
+
+            fullType.TryGetType(path, out var typeOuter);
+            if (!typeOuter.IsAggregate)
+            {
+                return typeOuter.Kind == kind;
+            }
+
+            var tree = typeOuter.TypeTree;
+            foreach (var typedName in fullType.GetNames(path))
+            {
+                if (typedName.Type.Kind == kind)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        public bool ContainsKindNested(DPath path, DKind kind)
+        {
+            AssertValid();
+            Contracts.Assert(kind >= DKind._Min && kind < DKind._Lim);
+
+            var fullType = this;
+            if (IsLazyType)
+            {
+                // This probably should throw (or be eliminated)
```

Comment:
Yeah. For Lazy Types, all these *Nested operations should only look at the top layer and not recurse, imo.

Reply:
We can then have runtime errors if something deeper than that violates the constraint we tried to enforce with this type function.

Reply:
Done